mev_inspect/arbitrages.py

```diff
@@ -23,70 +23,121 @@ def get_arbitrages(swaps: List[Swap]) -> List[Arbitrage]:
                 end_amount=end_amount,
                 profit_amount=profit_amount,
             )
+            all_arbitrages.append(arb)
 
-    return None
-
+    return all_arbitrages
 
-def _get_swap_from_address(
-    address: str, token_address: str, swaps: List[Swap]
-) -> Optional[Swap]:
-    for swap in swaps:
-        if swap.pool_address == address and swap.token_in_address == token_address:
-            return swap
 
-    return None
+def _get_all_start_end_swaps(swaps: List[Swap]) -> List[Tuple[Swap, Swap]]:
+    """
+    Gets the set of all possible opening and closing swap pairs in an arbitrage via
+    - swap[start].token_in == swap[end].token_out
+    - swap[start].token_in_amount < swap[end].token_out_amount (assumes profitable)
+    """
+    # TODO: Consider also filtering by start.from == end.to -- duplicate for now.
+    results: List[Tuple[Swap, Swap]] = []
+    for potential_start_swap in swaps:
+        for potential_end_swap in swaps:
+            # Checking by profitability can over-discover due to MEV-bot dust. Ex: 3 pool arb, second tx does not use all of output of first, we'd classify two start/end points. These have to be filtered separately.
+            if (
+                potential_start_swap.token_in_address
+                == potential_end_swap.token_out_address
+                and potential_start_swap.token_in_amount
+                < potential_end_swap.token_out_amount
+            ):
+                results.append((potential_start_swap, potential_end_swap))
+    return results
+
+
+def _get_all_routes(
+    start_swap: Swap, end_swap: Swap, other_swaps: List[Swap]
+) -> List[List[Swap]]:
+    """
+    Returns all routes (List[Swap]) from start to finish between a start_swap and an end_swap only accounting for token_address_in and token_address_out.
+    """
+    # If the path is complete, return
+    if start_swap.token_out_address == end_swap.token_in_address:
+        return [[start_swap, end_swap]]
+    elif len(other_swaps) == 0:
+        return []
+
+    # Collect all potential next steps, check if valid, recursively find routes from next_step to end_swap
+    routes: List[List[Swap]] = []
+    for potential_next_swap in other_swaps:
+        if start_swap.token_out_address == potential_next_swap.token_in_address:
+            remaining_swaps = [
+                swap for swap in other_swaps if swap != potential_next_swap
+            ]
+            next_swap_routes = _get_all_routes(
+                potential_next_swap, end_swap, remaining_swaps
+            )
+            if len(next_swap_routes) > 0:
+                for next_swap_route in next_swap_routes:
+                    next_swap_route.insert(0, start_swap)
+                    routes.append(next_swap_route)
+    return routes
+
+
+def _swaps_have_valid_order_wrt_trace_address(swaps: List[Swap]) -> bool:
+    """
+    Checks if a list of swaps have a potentially valid order exclusively based on `swap.trace_address`.
+    """
```

Comment:
If I'm reading right, valid means that the order in `swaps` is the same order as they occurred in the trace - is that right?

If so, because trace ordering is the natural ordering when comparing lists of ints, you could do something like this which I think more reads like that description:
```
swaps_trace_addresses = [swap.trace_address for swap in swaps]
ordered_trace_addresses = list(sorted(swaps_trace_addresses))

return swaps_trace_addresses == ordered_trace_addresses
```

Reply:
This is slightly less specific to allows for **some** out of order txs (this is from `test_arbitrages.py`). Basically it still allows for flash swaps. 

```
    # Block 12914944
    swaps = [
        create_generic_swap(trace_address=[0, 0, 2, 0, 2, 1]),
        create_generic_swap(trace_address=[0, 0, 2, 0]),
        create_generic_swap(trace_address=[0, 0]),
    ]
```

Regardless, I put a new proposal above that won't rely on this kind of logic.